src/tree/builder/TreeAdapter.js

```diff
@@ -0,0 +1,449 @@
+// Tree elements.
+import FormattingElement from "../structure/FormattingElement";
+import Heading from "../structure/nodes/Heading";
+import List from "../structure/nodes/List";
+import ListItem from "../structure/nodes/ListItem";
+import Paragraph from "../structure/nodes/Paragraph";
+import StructuredIrrelevant from "../structure/nodes/StructuredIrrelevant";
+import StructuredNode from "../structure/nodes/StructuredNode";
+// HTML classes.
+import { formattingElements, headings, irrelevantHtmlElements } from "./htmlClasses";
+
+/**
+ * An adapter to parse the HTML source code to a structured tree representation, to be used in further analysis,
+ * in combination with the `parse5` library.
+ *
+ * Implements the `parse5` library's `TreeAdapter` interface
+ * @see https://github.com/inikulin/parse5/blob/master/packages/parse5/docs/tree-adapter/interface.md
+ *
+ * @memberOf module:tree/builder
+ *
+ * @private
+ */
+class TreeAdapter {
+	// Creation of nodes and other tree elements.
+
+	/**
+	 * Creates a new element to be put in the tree.
+	 *
+	 * @param {string} tag									The HTML tag.
+	 * @param {string} namespace							The XML namespace (e.g. "http://www.w3.org/1999/xhtml" for HTML).
+	 * @param {Array<{ name: string, value: string }>} attributes	The attributes of this element.
+	 *
+	 * @returns {module:tree/structure.Node|module:tree/structure.FormattingElement} The new element.
+	 */
+	createElement( tag, namespace, attributes ) {
+		let node;
+
+		if ( irrelevantHtmlElements.includes( tag ) ) {
+			// Irrelevant for analysis (e.g. `script`, `style`).
+			node = new StructuredIrrelevant( tag );
+		} else if ( formattingElements.includes( tag ) ) {
+			// Formatting element.
+			const parsedAttributes = TreeAdapter._parseAttributes( attributes );
+			node = new FormattingElement( tag, parsedAttributes );
+		} else {
+			// Paragraphs, Headers, Lists, ListItems and other nodes.
+			node = this._parseNode( tag );
+		}
+
+		node.tagName = tag;
+		node.namespace = namespace;
+		node.parent = null;
+
+		return node;
+	}
+
+	/**
+	 * Makes a new node to add to the tree, based on the given HTML-tag.
+	 *
+	 * @param {string} tag		The HTML-tag of the element to add.
+	 *
+	 * @returns {module:tree/structure.Node} The node to add to the tree.
+	 *
+	 * @private
+	 */
+	_parseNode( tag ) {
+		if ( headings.includes( tag ) ) {
+			// Heading.
+			return new Heading( parseInt( tag[ 1 ], 10 ) );
+		} else if ( tag === "p" ) {
+			// Paragraph.
+			return new Paragraph( tag );
+		} else if ( tag === "ol" || tag === "ul" ) {
```

Comment:
Do you mean `dl` lists? 

They are somewhat different than ordered and unordered lists in that you must specify individual keys, using `dt` tags. Which the current code does not support yet.

(In the case of ordered and unordered lists, the keys can be deduced by the kind of list instead (e.g. "1, 2, 3, ..." in the case of ordered, bullet points in the case of unordered lists).)

Since this PR already adds huge chunk of new functionality, I would make an issue and add support for `dl` lists later on instead.

Reply:
Made a new issue: #2094.